source/linker.cpp

```diff
@@ -0,0 +1,1098 @@
+                                 std::vector<std::vector<uint32_t>> &binaries,
+                                 size_t num_binaries, uint32_t max_index,
+                                 uint32_t version,
+                                 spv_binary_t& merged_spvbinary,
+                                 bool create_library) {
+  spv_position_t position = {};
+
+  std::vector<uint32_t> indices(num_binaries, 5u);
+
+  std::unordered_map<spv::Id, const char *> exports;
+  std::unordered_map<spv::Id, const char *> imports;
+  std::unordered_map<spv::Id, spv::Id> link_table;
+  std::vector<spv::Id> builtins;
+
+  uint32_t* merged_binary = merged_spvbinary.code;
+  size_t& merged_length = merged_spvbinary.wordCount;
+
+  // Generate new header
+  SetWord(merged_binary, 0u, spv::MagicNumber);
+  SetWord(merged_binary, 1u, version);
+  SetWord(merged_binary, 2u, 0u);
+  SetWord(merged_binary, 3u, max_index);
+  SetWord(merged_binary, 4u, 0u);
+  merged_length = 5u;
+
+  const auto CopySections = [&binaries,&indices,num_binaries,
+                             &merged_binary,&merged_length](const std::function<bool (spv::Op)>& check_section) {
+    for (uint32_t i = 0u; i < num_binaries; ++i) {
+      const uint32_t* const binary = binaries[i].data();
+      const size_t length = binaries[i].size();
+      const uint32_t index = indices[i];
+      uint32_t j = index;
+      uint32_t prev_j = j;
+
+      while (j < length) {
+        uint32_t word_count;
+        const spv::Op opcode = ParseOpcode(binary, j, word_count);
+
+        prev_j = j;
+
+        if (!check_section(opcode))
+          break;
+
+        j += word_count;
+      }
+      if (j == length)
+        prev_j = j;
+
+      std::memcpy(merged_binary + merged_length, binary + index,
+                  (prev_j - index) * sizeof(spirv_word));
+      merged_length += prev_j - index;
+      indices[i] = j;
+    }
+  };
+
+  // Merge all OpCapability
+  std::unordered_set<uint32_t> capabilities;
+  for (uint32_t i = 0u; i < num_binaries; ++i) {
+    const uint32_t* const binary = binaries[i].data();
+    const size_t length = binaries[i].size();
+    const uint32_t index = indices[i];
+    uint32_t j = index;
+    uint32_t prev_j = j;
+
+    while (j < length) {
+      uint32_t word_count;
+      const spv::Op opcode = ParseOpcode(binary, j, word_count);
+
+      prev_j = j;
+
+      if (!IsCapabilityOpcode(opcode))
+        break;
+
+      capabilities.insert(GetWord(binary, j + 1u));
+
+      j += word_count;
+    }
+
+    indices[i] = j;
+  }
+  for (auto i : capabilities) {
+    merged_binary[merged_length++] = static_cast<uint32_t>(spv::Op::OpCapability) | (2u << spv::WordCountShift);
+    merged_binary[merged_length++] = i;
+  }
+
+  // Copy OpExtensions
+  CopySections(IsExtensionOpcode);
+
+  // Copy OpExtInstImport
+  CopySections(IsExtInstImportOpcode);
+
+  // Merge OpMemoryModel
+  spv::AddressingModel addressing_model, addressing_model_tmp;
```

Comment:
I will do that. I should probably use `spv::AddressModel::Max` and the memory model equivalent for the initialisation, as I don’t see what could be a good default value. And if after processing the binaries we still have the default values, then something wrong happened.